### PR TITLE
feat: customize variable qualities

### PR DESCRIPTION
**Pull request by ashutoshkrris**
## Description

Attempts to fix: #100 

Before this PR, users were not able to add `variable_qualities` option while creating srcset.

After this PR, users can optionally pass a `variable_qualities` dictionary in `options` while creating srcset.

## Steps to test
* Before this PR:
```py
>>> from imgix import UrlBuilder
>>> client = UrlBuilder("sdk-test.imgix.net")
>>> client.create_srcset("image.jpg", {"w": 100}, {"variable_qualities": {1: 45, 2: 30, 3: 20, 4: 15, 5: 10}})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: UrlBuilder.create_srcset() takes from 2 to 3 positional arguments but 4 were given
```
Notice that the `create_srcset()` method throws TypeError when `variable_qualities` is passed.

* After this PR:
```py
>>> from imgix import UrlBuilder
>>> client = UrlBuilder("sdk-test.imgix.net")
>>> client.create_srcset("image.jpg", {"w": 100}, {"variable_qualities": {1: 45, 2: 30, 3: 20, 4: 15, 5: 10}})
'https://sdk-test.imgix.net/image.jpg?dpr=1&q=45&w=100 1x,\nhttps://sdk-test.imgix.net/image.jpg?dpr=2&q=30&w=100 2x,\nhttps://sdk-test.imgix.net/image.jpg?dpr=3&q=20&w=100 3x,\nhttps://sdk-test.imgix.net/image.jpg?dpr=4&&q=15&w=100 4x,\nhttps://sdk-test.imgix.net/image.jpg?dpr=5&q=10&w=100 5x'
```
Notice that the output URLs have the same DPR and quality value that we have passed in the `create_srcset` function.

## Checklist

- [x] Read the [contributing guidelines](CONTRIBUTING.md).
- [x] Each commit follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) spec format.
- [x] Update the readme (if applicable).
- [ ] Update or add any necessary API documentation (if applicable)
- [x] All existing unit tests are still passing (if applicable).

- [x] Add some [steps](#steps-to-test) so we can test your bug fix or feature (if applicable).
- [x] Add new passing unit tests to cover the code introduced by your PR (if applicable).
- [ ] Any breaking changes are specified on the commit on which they are introduced with `BREAKING CHANGE` in the body of the commit.
- [ ] If this is a big feature with breaking changes, consider opening an issue to discuss first. This is completely up to you, but please keep in mind that your PR might not be accepted.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -170,6 +171,25 @@ https://demo.imgix.net/image.jpg?w=100&dpr=4&q=23 4x,
...
 
+#### Variable Qualities
+
+By default, this library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset.
+
+To customize variable qualities, you can pass a `variable_qualities` dictionary in the `options` while creating srcset as below:
+```py
+>>> from imgix import UrlBuilder
+>>> client = UrlBuilder("sdk-test.imgix.net")
+>>> client.create_srcset("image.jpg", {"w": 100}, {"variable_qualities": {1: 45, 2: 30, 3: 20, 4: 15, 5: 10}})
```
Comment (sherwinski) [inline README.md:182]: For consistency, we should remove the prompts usually seen when using the python interpreter.

```suggestion
from imgix import UrlBuilder
client = UrlBuilder("sdk-test.imgix.net")
client.create_srcset("image.jpg", {"w": 100}, {"variable_qualities": {1: 45, 2: 30, 3: 20, 4: 15, 5: 10}})
```

File: README.md [hunk 2]
```diff
@@ -170,6 +171,25 @@ https://demo.imgix.net/image.jpg?w=100&dpr=4&q=23 4x,
 https://demo.imgix.net/image.jpg?w=100&dpr=5&q=20 5x
 ```
 
+#### Variable Qualities
+
+By default, this library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset.
+
+To customize variable qualities, you can pass a `variable_qualities` dictionary in the `options` while creating srcset as below:
+```py
```
Comment (sherwinski) [inline README.md:179]:
```suggestion
```python
```

File: imgix/validators.py [hunk 3]
```diff
@@ -174,11 +180,31 @@ def validate_widths(width_values):
...
+def validate_variable_qualities(variable_qualities):
+    """Validate a list of `variable_qualities`.
+
+    This function checks whether `variable_qualities` is a dictionary.
+
+
+    Parameters
+    ----------
+    variable_qualities : dict
+        A dictionary of variable qualities values.
```
Comment (sherwinski) [inline imgix/validators.py:202]:
```suggestion
        A dictionary of variable quality values.
```

File: imgix/urlbuilder.py [hunk 4]
```diff
@@ -322,12 +328,23 @@ def _build_srcset_DPR(
...
+
         for dpr in targets:
             srcset_params["dpr"] = dpr
 
             if not disable_variable_quality:
-                quality = params.get("q", DPR_QUALITIES[dpr])
-                srcset_params["q"] = quality
+                quality = (
+                    params.get("q", None)
+                    or qualities.get("dpr")
```
Comment (sherwinski) [inline imgix/urlbuilder.py:343]: Could you elaborate on the purpose of checking for this? I tried testing the following and there appears to have no effect:

```python
client.create_srcset("image.jpg", {"w": 100, "dpr":1}, {"variable_qualities": {1: 100, 2: 90, 3: 80}})
```

File: tests/test_srcset.py [hunk 5]
```diff
@@ -96,6 +96,24 @@ def test_create_srcset_with_widths():
     assert expected == actual
 
 
+def test_create_srcset_with_variable_qualities():
```
Comment (sherwinski) [inline tests/test_srcset.py:99]: This test is good but I think we should also allow the user to have less than five srcset ratio descriptors if that's what they explicitly call for. Could we add a test for the following and ensure that it works?

```python
 ub = imgix.UrlBuilder(DOMAIN, include_library_param=False)
client.create_srcset("image.jpg", { "w": 100 }, { "variable_qualities": {1: 100, 2: 90, 3: 80} })

# https://testing.imgix.net/image.jpg?dpr=1&q=100&w=100 1x,
# https://testing.imgix.net/image.jpg?dpr=2&q=90&w=100 2x,
# https://testing.imgix.net/image.jpg?dpr=3&q=80&w=100 3x,
```

File: README.md [hunk 6]
```diff
@@ -170,6 +171,25 @@ https://demo.imgix.net/image.jpg?w=100&dpr=4&q=23 4x,
...
+>>> from imgix import UrlBuilder
+>>> client = UrlBuilder("sdk-test.imgix.net")
```
Comment (sherwinski) [inline README.md:181]: Let's keep the image domain/path consistent with what's already demo'd in the documentation, e.g. `https://demo.imgix.net/image.jpg?...`

We'll need to change this in the [outputted results](https://github.com/imgix/imgix-python/pull/102/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R186-R190) section as well.

```suggestion
builder = UrlBuilder("demo.imgix.net")
```

**Review by sherwinski — CHANGES_REQUESTED** [review]
Hey @ashutoshkrris and thank you for the PR 👋 
I had a couple of points of feedback and one open question if you don't mind taking a look and responding.

File: tests/test_srcset.py (hunk 5 above)
Comment (ashutoshkrris) [reply tests/test_srcset.py:99]: Doesn't this relate to issue #101?

Comment (sherwinski) [reply tests/test_srcset.py:99]: I'm not sure I understand your question @ashutoshkrris.
The current test case we have is good but I think we should also write a test, and implement any changes for, the [use case](https://github.com/imgix/imgix-python/pull/102#discussion_r1018671654) I mentioned previously.

**Comment by sherwinski** [comment]
Hey @ashutoshkrris, just checking in here. Once we get the feedback here incorporated into this PR, we should be good to move forward with merging. Please let me know if you have any questions.

**Comment by ashutoshkrris** [comment]
Hi @sherwinski 
I have added the feature to allow less than five variable qualities too, please check.

File: imgix/urlbuilder.py (hunk 4 above)
Comment (sherwinski) [reply imgix/urlbuilder.py:344]: Ok that makes sense. We'll have to check back in on this specific condition another time.

**Comment by sherwinski** [comment]
@ashutoshkrris Thanks for taking the time to address my feedback and appreciate the effort you put into this PR 🚀

**Pull request merged**